pisa/isopolygons.py

```diff
@@ -0,0 +1,313 @@
+
+        Returns:
+            nodes_gdf: a GeoSeries of the nodes with their osmid and geometry.
+            edges_gdf: a GeoSeries of the geometry of the edges.
+
+
+        If an edge (u,v) doesn't have geometry data in G, edges_gdf contains
+        a straight line from u to v.
+
+        If no edges are found (for example, if all other nodes are too far away from center_node),
+        edges_gdf is an empty dataframe.
+
+        """
+        subgraph = nx.ego_graph(
+            road_network, center_node, radius=distance_value, distance=distance_type
+        )
+
+        try:
+            nodes_gdf, edges_gdf = ox.graph_to_gdfs(subgraph)
+            return (
+                nodes_gdf.loc[:, "geometry"],
+                edges_gdf.loc[:, "geometry"].reset_index(),
+            )
+
+        except ValueError:  # if no edges are found
+            nodes_gdf = ox.graph_to_gdfs(subgraph, edges=False)
+            return (
+                nodes_gdf.loc[:, "geometry"],
+                gpd.GeoSeries([], name="geometry", crs=nodes_gdf.crs),
+            )
+
+
+class OsmIsopolygonCalculatorAlternative(IsopolygonCalculator):
+
+    def __init__(
+        self,
+        facilities_lon_lat: DataFrame,
+        distance_type: str,
+        distance_values: list[int],
+        road_network: MultiDiGraph,
+        buffer: float = 50,  # in meters
+    ):
+        super().__init__(facilities_lon_lat, distance_type, distance_values)
+        self.road_network = road_network
+        self.buffer = buffer
+
+        # For each facility, find the nearest node in the road network
+        self.nearest_nodes = ox.distance.nearest_nodes(
+            G=self.road_network,
+            X=self.facilities_longitude_array,
+            Y=self.facilities_latitude_array,
+        )
+
+    def calculate_isopolygons(self) -> DataFrame:
+        """
+        Calculate isopolygons for each facility at different distances (distance_values).
+
+        An isopolygon represents the area that can be reached within a specific distance/time from a facility
+        using the road network.
+
+        """
+
+        isopolygons = pd.DataFrame()
+
+        for distance_value in self.distance_values:
+
+            for road_node in self.nearest_nodes:
+
+                # Get the nodes and edges within distance_value from road_node
+                skeleton = self._get_skeleton_nodes_and_edges(
+                    self.road_network, road_node, distance_value, self.distance_type
+                )
+
+                # "Inflate" the skeleton into a Polygon
+                isopolygon = ox.utils_geo.buffer_geometry(
+                    geom=skeleton, dist=self.buffer
+                )
+
+                # Add the isopolygon to the isopolygons DataFrame at row road_node and column distance_value
+                isopolygons.at[road_node, "ID_" + str(distance_value)] = isopolygon
+
+        return isopolygons
+
+    @staticmethod
+    def _get_skeleton_nodes_and_edges(
```

Comment:
I like this approach. the output is already a unified geometry which is easier to understand and handle.